src/main/java/net/snowflake/client/jdbc/cloud/storage/GcmEncryptionProvider.java

```diff
@@ -0,0 +1,245 @@
+
+    byte[] kek = base64Decoder.decode(encMat.getQueryStageMasterKey());
+    int keySize = kek.length;
+    byte[] keyBytes = new byte[keySize];
+    byte[] dataIvBytes = new byte[blockSize];
+    byte[] keyIvBytes = new byte[blockSize];
+    initRandomIvsAndFileKey(dataIvBytes, keyIvBytes, keyBytes);
+    byte[] encryptedKey = encryptKey(kek, keyBytes, keyIvBytes, keyAad);
+    CipherInputStream cis = encryptContent(src, keyBytes, dataIvBytes, dataAad);
+    addEncryptionMetadataToStorageClient(
+        meta,
+        originalContentLength,
+        encMat,
+        client,
+        keySize,
+        encryptedKey,
+        dataIvBytes,
+        keyIvBytes,
+        keyAad,
+        dataAad);
+    return cis;
+  }
+
+  private static void initRandomIvsAndFileKey(
+      byte[] cekIvData, byte[] kekIvData, byte[] fileKeyBytes) {
+    random.nextBytes(cekIvData);
+    random.nextBytes(kekIvData);
+    random.nextBytes(fileKeyBytes);
+  }
+
+  private static byte[] encryptKey(byte[] kekBytes, byte[] keyBytes, byte[] keyIvData, byte[] aad)
+      throws InvalidKeyException, InvalidAlgorithmParameterException, IllegalBlockSizeException,
+          BadPaddingException, NoSuchPaddingException, NoSuchAlgorithmException {
+    SecretKey kek = new SecretKeySpec(kekBytes, 0, kekBytes.length, AES);
+    GCMParameterSpec gcmParameterSpec = new GCMParameterSpec(TAG_LENGTH, keyIvData);
+    Cipher keyCipher = Cipher.getInstance(KEY_CIPHER);
+    keyCipher.init(Cipher.ENCRYPT_MODE, kek, gcmParameterSpec);
+    if (aad != null) {
+      keyCipher.updateAAD(aad);
+    }
+    return keyCipher.doFinal(keyBytes);
+  }
+
+  private static CipherInputStream encryptContent(
+      InputStream src, byte[] keyBytes, byte[] dataIvBytes, byte[] aad)
+      throws InvalidKeyException, InvalidAlgorithmParameterException, NoSuchPaddingException,
+          NoSuchAlgorithmException {
+    SecretKey fileKey = new SecretKeySpec(keyBytes, 0, keyBytes.length, AES);
+    GCMParameterSpec gcmParameterSpec = new GCMParameterSpec(TAG_LENGTH, dataIvBytes);
+    Cipher fileCipher = Cipher.getInstance(FILE_CIPHER);
+    fileCipher.init(Cipher.ENCRYPT_MODE, fileKey, gcmParameterSpec);
+    if (aad != null) {
+      fileCipher.updateAAD(aad);
+    }
+    return new CipherInputStream(src, fileCipher);
+  }
+
+  private static void addEncryptionMetadataToStorageClient(
+      StorageObjectMetadata meta,
+      long originalContentLength,
+      RemoteStoreFileEncryptionMaterial encMat,
+      SnowflakeStorageClient client,
+      int keySize,
+      byte[] encryptedKey,
+      byte[] dataIvData,
+      byte[] keyIvData,
+      byte[] keyAad,
+      byte[] dataAad) {
+    // Comment from EncryptionProvider:
+    // Round up length to next multiple of the block size
+    // Sizes that are multiples of the block size need to be padded to next
+    // multiple
+    // TODO GCM SNOW-1431870 do we really need it, if padding is not used?
```

Comment:
I don't really understand this TODO. Do you mean "are NOT multiples"?

We should not need to pad the message for GCM.